cep-recipe-jinja.md

```diff
@@ -0,0 +1,527 @@
+
+The function should evaluate to `{<lang>_stdlib}_{target_platform}
+<lang>_stdlib_version`.
+
+### The `cdt` function
+
+CDT stands for "core dependency tree" packages. These are typically repackaged
+from a Linux distribution.
+
+The function expands to the following:
+
+- package-name-<cdt_name>-<cdt_arch>
+
+Where `cdt_name` and `cdt_arch` are loaded from the variant config. If they are
+undefined in the variant configuration, an error is raised. There are no default
+values for `cdt_name` and `cdt_arch`.
+
+### The pin functions
+
+The new recipe format has two `pin` expressions:
+
+- `pin_compatible`
+- `pin_subpackage`
+
+Both follow the same "pinning" mechanism as described next and have the same
+arguments.
+
+#### Pin definition
+
+A pin has the following arguments:
+
+- `package_name`, positional, required: The name of the package to pin.
+- `lower_bound`, defaults to `x.x.x.x.x.x`: the lower bound, either as a version
+  or as a "pin expression", or `None`
+- `upper_bound`, defaults to `x`: the upper bound, either as a version or as a
+  "pin expression" or `None`
+- `exact`: a boolean that specifies whether the pin should be exact. It defaults
+  to `False`. If `exact` is `True`, the `lower_bound` and `upper_bound` are
+  irrelevant and should not be set.
+  An exact pin must pin with the full version and build string (to a
+  single package), e.g. `==version=build`.
+
+##### Pin expressions
+
+A pin expression is a string that contains only `x` and `.` characters. The
+number of `x` characters in the expression determines the number of segments
+that are used from the version.
+
+A pin expression of `x.x` applied to a version like `1.2.3` would yield `1.2`.
+The epoch and local version parts are left untouched by the pin expression:
+`1!1.2.3+local` with a `x.x` pin expression would yield `1!1.2+local`.
+
+The version used in the pin expression computation must _always_ be the version
+that was determined during the run of the recipe (irrespective of setting the
+lower bound to an explicit version).
+
+##### Upper bound pin computation
+
+When a pin expression is used for the upper bound, the last segment of the
+version must be incremented, and the local version part must be removed.
+
+- If the last segment is a letter, the number should be incremented and the
+  letter set to `a`, e.g. `9d` with a `x` pin expression results in `<10a`.
+- If the last segment is a number, the number should be incremented and `.0a0`
+  should be appended to prevent any alpha versions from being selected. For
+  example: `1.2.3` with a `x.x` pin expression should result in `<1.3.0a0`.
+- The epoch is left untouched by the `max_pin` (or `min_pin`). If the epoch is
+  set, it will be included in the final version. E.g. `1!1.2.3` with a
+  `max_pin='x.x'` will result in `<1!1.3.0a0`.
+- When bumping the version with a `max_pin` the local version part is removed.
+  For example, `1.2.3+local` with a `max_pin='x.x'` will result in `<1.3.0a0`.
```

Comment:
Whereas this is specifying pins. Should we consolidate to one terminology or should we define both? Are there cases we need both?